### PR TITLE
Pass --timeout flag to pyperf

### PR DESCRIPTION
**Pull request by diegorusso**
Add the `--timeout` flag to the `run` command. By default there is no timeout and user needs to specify it if they want it. I wanted to avoid to change the current behaviour.
This addressed the issue https://github.com/python/pyperformance/issues/353

If you run the below command passing the timeout of 1 second, dask will fail and it will be reported at the end.

```
$ pyperformance run --debug-single-value -b float,go,mako,dask --timeout 1
...
...
Performance version: 1.11.0
Python version: 3.12.3 (64-bit)
Report on Linux-6.8.0-44-generic-aarch64-with-glibc2.39
Number of logical CPUs: 8
Start date: 2024-09-19 10:11:20.472768
End date: 2024-09-19 10:11:21.843698

### float ###
85.8 ms

### go ###
119 ms

### mako ###
9.77 ms

1 benchmarks failed:
- dask (Timed out after 1 seconds)
```

### CODE REVIEW

**Comment by nineteendo** [comment]
Could you document `--min-time` in a different pull request?

File: doc/usage.rst [hunk 1]
```diff
@@ -101,9 +101,10 @@ Usage::
 
   pyperformance run [-h] [-r] [-f] [--debug-single-value] [-v] [-m]
                        [--affinity CPU_LIST] [-o FILENAME]
-                       [--append FILENAME] [--manifest MANIFEST]
-                       [-b BM_LIST] [--inherit-environ VAR_LIST]
-                       [-p PYTHON]
+                       [--append FILENAME] [--min-time MIN_TIME]
+                       [--same-loops SAME_LOOPS] [--timeout TIMEOUT]
```
Comment (nineteendo) [inline doc/usage.rst:105]: Could you document `--min-time` and `--same-loops` in a different pull request?

File: pyperformance/commands.py [hunk 2]
```diff
@@ -191,8 +191,8 @@ def cmd_run(options, benchmarks):
 
     if errors:
         print("%s benchmarks failed:" % len(errors))
-        for name in errors:
-            print("- %s" % name)
+        for name, reason in errors:
+            print("- %s (%s)" % (name, reason))
```
Comment (nineteendo) [inline pyperformance/commands.py:195]: Should this be in a separate pull request?

File: doc/usage.rst (hunk 1 above)
Comment (diegorusso) [reply doc/usage.rst:105]: `--same-loops` has just moved from the end of the list to the right place. it was misplaced in the first place.
`--min-time` was never added here but it's not new.

Shall I put both of them in a PR?

**Review by vstinner — COMMENTED** [review]
pyperf main process does basically nothing, it just waits until worker processes complete. It may be safer to implement the timeout in pyperf, to kill both processes, the main one and the current worker process.

If you implement the timeout in pyperformance, you only see the main process and you cannot (easily) kill the worker process.

File: doc/usage.rst (hunk 1 above)
Comment (nineteendo) [reply doc/usage.rst:105]: Moving it to a different place is probably fine, but updating the usage should probably be in a separate PR.

File: pyperformance/commands.py (hunk 2 above)
Comment (diegorusso) [reply pyperformance/commands.py:195]: The timeout implementation introduces a new way that a benchmark can fail. This is really to differentiate the ways a benchmarks can terminate hence I thought to be explicit and communicate it back to the user what the reason of the termination is.
If you want I can split it in an extra PR.

Comment (nineteendo) [reply pyperformance/commands.py:195]: It's not up to me to decide that. I'm not a maintainer. :)

File: doc/usage.rst (hunk 1 above)
Comment (mdboom) [reply doc/usage.rst:105]: I implemented `--same-loops` a few months ago -- it's documented in here and in the usage.rst.  Was there something else you are looking for?  I'll volunteer to submit if so (in a separate PR).

File: doc/usage.rst [hunk 3]
```diff
@@ -124,10 +125,17 @@ options::
                         baseline_python, not changed_python.
   --append FILENAME     Add runs to an existing file, or create it if
                         it doesn't exist
+  --min-time MIN_TIME   Minimum duration in seconds of a single value, used
+                        to calibrate the number of loops
+  --same-loops SAME_LOOPS
+                        Use the same number of loops as a previous run
+                        (i.e., don't recalibrate). Should be a path to a
+                        .json file from a previous run.
+  --timeout TIMEOUT     Timeout for a benchmark run (default: disabled)
```
Comment (mdboom) [inline doc/usage.rst:134]: What's the unit here?  Seconds?

File: pyperformance/commands.py (hunk 2 above)
Comment (mdboom) [reply pyperformance/commands.py:195]: This change to show the reason makes sense to me.

File: pyperformance/tests/__init__.py [hunk 4]
```diff
@@ -39,6 +39,10 @@ def run_cmd(cmd, *args, capture=None, onfail='exit', verbose=True):
 
     if verbose:
         print(f"(tests) Execute: {argv_str}", flush=True)
+
+    if timeout:
+        kwargs['timeout'] = 60
```
Comment (mdboom) [inline pyperformance/tests/__init__.py:44]: Should we add a test that deliberately times out to confirm it's working?

**Review by mdboom — COMMENTED** [review]
I may have sent you forth with bad information by suggesting that this would be easier in `pyperformance`, and maybe @vstinner is right that this would be better handled in `pyperf`.

We'd still need a patch to `pyperformance` to pass the `--timeout` value down to pyperf -- 80% of this PR is handling commandline arguments as it is, so not wasted effort.

File: doc/usage.rst (hunk 3 above)
Comment (diegorusso) [reply doc/usage.rst:134]: yes, they are seconds. I think I should be explicit. I'll address it tomorrow.

File: doc/usage.rst (hunk 1 above)
Comment (diegorusso) [reply doc/usage.rst:105]: The changes in usage.rst are there to reflect the actual output of pyperformance. In the PR https://github.com/python/pyperformance/pull/327 implemented the same loops option but it has been forgotten to be added here.

File: doc/usage.rst [hunk 5]
```diff
@@ -124,10 +125,17 @@ options::
                         baseline_python, not changed_python.
   --append FILENAME     Add runs to an existing file, or create it if
                         it doesn't exist
+  --min-time MIN_TIME   Minimum duration in seconds of a single value, used
+                        to calibrate the number of loops
+  --same-loops SAME_LOOPS
```
Comment (diegorusso) [inline doc/usage.rst:130]: Similarly to the comment above (re. same loops), argparse positions the option here and not at the end like used to be. This because it is an option of the run command.
--manifest, -b and -p flags are option of pyperformance hence they are placed at the end.

File: pyperformance/tests/__init__.py (hunk 4 above)
Comment (diegorusso) [reply pyperformance/tests/__init__.py:44]: yes, it makes sense. I'll add it tomorrow.

**Comment by diegorusso** [comment]
> I may have sent you forth with bad information by suggesting that this would be easier in `pyperformance`, and maybe @vstinner is right that this would be better handled in `pyperf`.
> 
> We'd still need a patch to `pyperformance` to pass the `--timeout` value down to pyperf -- 80% of this PR is handling commandline arguments as it is, so not wasted effort.

Don't worry, that's the purpose of the review. I leave this PR open for now and implement the timeout in pyperf. Once that one is in, I'll fix this up to make use of the timeout.

File: doc/usage.rst (hunk 5 above)
Comment (diegorusso) [reply doc/usage.rst]: Dropped from this PR

File: doc/usage.rst (hunk 3 above)
Comment (diegorusso) [reply doc/usage.rst]: Added.

**Comment by diegorusso** [comment]
I've addressed feedbacks on this PR and also raised a new one in pyperf: https://github.com/psf/pyperf/pull/205

File: pyperformance/tests/__init__.py (hunk 4 above)
Comment (diegorusso) [reply pyperformance/tests/__init__.py]: As the timeout is implemented in pyperf now (https://github.com/psf/pyperf/pull/205) I created a test in that PR. pyperformance now just passes the info to pyperf.
Would you like to have a test that checks the timeout is actually passed to pyperf?

Comment (mdboom) [reply pyperformance/tests/__init__.py]: My comment was for earlier when timeout was handled here. I'm fine with not having a timeout test in pyperformance now that the actual implementation is in pyperf.

**Comment by diegorusso** [comment]
Just to let you know that the [pyperf PR](https://github.com/psf/pyperf/pull/205) has been merged (thanks @vstinner) so we have the timeout functionality there.

**Comment by vstinner** [comment]
> Just to let you know that the https://github.com/psf/pyperf/pull/205 has been merged (thanks @vstinner) so we have the timeout functionality there.

You need a pyperf release first.

**Comment by mdboom** [comment]
> > Just to let you know that the [psf/pyperf#205](https://github.com/psf/pyperf/pull/205) has been merged (thanks @vstinner) so we have the timeout functionality there.
> 
> You need a pyperf release first.

I started one [here](https://github.com/psf/pyperf/pull/206)

**Comment by vstinner** [comment]
Your PR should update pyperf to 2.8.

**Comment by diegorusso** [comment]
> Your PR should update pyperf to 2.8.

I see that pyperf is unpinned so in theory I should not change anything in there. I have done a fresh install of pyperformance and in fact it installs pyperf 2.8

**Comment by mdboom** [comment]
> > Your PR should update pyperf to 2.8.
> 
> I see that pyperf is unpinned so in theory I should not change anything in there. I have done a fresh install of pyperformance and in fact it installs pyperf 2.8

I think you still need to update it [here](https://github.com/python/pyperformance/blob/main/pyperformance/requirements/requirements.txt#L13), which is the version that will actually run in the venvs.

**Comment by diegorusso** [comment]
> > > Your PR should update pyperf to 2.8.
> > 
> > 
> > I see that pyperf is unpinned so in theory I should not change anything in there. I have done a fresh install of pyperformance and in fact it installs pyperf 2.8
> 
> I think you still need to update it [here](https://github.com/python/pyperformance/blob/main/pyperformance/requirements/requirements.txt#L13), which is the version that will actually run in the venvs.

Yes, you are right. I've just found it! :)

**Comment by diegorusso** [comment]
I'm currently running the tests because there is a small change in the way pyperf reports unit measure: https://github.com/psf/pyperf/commit/432c419cff97804de8c780e82d2a8562e977275a

**Review by vstinner — APPROVED** [review]
LGTM

**Review by mdboom — APPROVED** [review]
Thanks!
